query/grouped_result.go

```diff
@@ -0,0 +1,99 @@
+package query
+
+import "github.com/xichen2020/eventdb/document/field"
+
+// TransformFn transforms a list of raw calculated values into a final calculated value.
+// For example, computation of `Avg` should be transformed into computation of `Sum` and
+// `Count`, and once the results are gathered, the final `Avg` value can be computed as
+// `Sum` / `Count`.
+type TransformFn func(rawValues []float64) float64
+
+// CalculatedValue represents a calculated value. In simple cases this is simply a single
+// numeric value. However, there are also cases where calculation of a value requires
+// calculations of other intermediate values, which eventually get transformed into the
+// final calculated value (e.g., the average value of a field).
+type CalculatedValue struct {
+	Value       float64
+	Values      []float64
+	TransformFn TransformFn
+}
+
+// ResultGroup is a result group.
+type ResultGroup struct {
+	// Group key corresponding to the `GroupBy` fields in the query.
+	Key []field.ValueUnion
+
+	// Field values to order the groups by.
+	OrderByValues []field.ValueUnion
+
+	// A list of calculated values for a result group.
+	CalculatedValues []CalculatedValue
+}
+
+// GroupedResults is a collection of result groups.
```

Comment:
Q: Does every segment within the bounds of the query return a `ResultGroup`? And `GroupedResults` is the collection of all results from every relevant segment.

Reply:
Each segment returns a set of `ResultGroups`, and the `GroupedResults` merges the results from different segments and different shards.